nexus/src/db/datastore.rs

```diff
@@ -1944,6 +1946,148 @@ impl DataStore {
             .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
     }
 
+    /// Update a network interface associated with a given instance.
+    pub async fn instance_update_network_interface(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+        authz_interface: &authz::NetworkInterface,
+        updates: NetworkInterfaceUpdate,
+    ) -> UpdateResult<NetworkInterface> {
+        use crate::db::schema::network_interface::dsl;
+
+        // This database operation is surprisingly subtle, and I don't believe
+        // it's possible to express in a single query.
```

Comment:
We really don't need to make this change now, but I'm pretty sure it would be possible.

Here's the case where `updates.make_primary == true`:

```sql
WITH
    /* Get instance state */
    instance as (
      SELECT * FROM instances
      WHERE id = <INSTANCE_ID>
        AND time_deleted IS NULL
    ),
    /* Get the primary interface */
    primary_interface as (
      SELECT * FROM network_interfaces
      WHERE instance_id = <INSTANCE_ID>
        AND is_primary
        AND time_deleted IS NULL
    ),
    /* Decide whether or not to do any updates. */
    do_update AS (
      SELECT IF(
         "Stopped" in (SELECT state from instance), TRUE, FALSE
      ),
    ),
    /* If the target and primary are different, convert primary to secondary */
    maybe_demoted_primary as (
      UPDATE network_interfaces SET is_primary = false
      WHERE id IN (SELECT id from primary_interface)
        AND time_deleted IS NULL
        AND (SELECT id from primary_interface) != <INTERFACE_ID>
        AND (SELECT * FROM do_update)
    ),
    /* Update the primary interface */
    updated_primary as (
      UPDATE network_interfaces SET <UPDATES>
      WHERE id = <INTERFACE_ID>
        AND time_deleted IS NULL
        AND (SELECT * FROM do_update)
      RETURNING *
    )
SELECT * FROM updated_primary
```

And here's the case where `updates.make_primary == false`:

```sql

WITH
    /* Get instance state */
    instance as (
      SELECT * FROM instances
      WHERE id = <INSTANCE_ID>
        AND time_deleted IS NULL
    ),
    /* Decide whether or not to do any updates. */
    do_update AS (
      SELECT IF(
         "Stopped" in (SELECT state from instance), TRUE, FALSE
      ),
    ),
    /* Update the primary interface */
    updated_primary as (
      UPDATE network_interfaces SET <UPDATES>
      WHERE id = <INTERFACE_ID>
        AND time_deleted IS NULL
        AND (SELECT * FROM do_update)
      RETURNING *
    )
SELECT * FROM updated_primary
```

Reply:
Nice! I swear I tried this early on, and then convinced myself that we couldn't have multiple update statements in a query. I'll give this a quick whirl. I still have the WIP branch with a CTE-based approach, so I might switch to that if this pans out. Thanks!

Reply:
So, I think this approach will work (thanks @smklein!), but it looks a bit risky to me. In particular, this query would stray perilously close to a [known bug](https://github.com/cockroachdb/cockroach/issues/70731) and [related settings](https://github.com/cockroachdb/cockroach/issues/76261) in CRDB. Specifically, it seems possible to corrupt a table when multiple modification statements touch the same rows. In this query, we'd be modifying _different_ rows, since this only hits more than one row if the primary and secondary NICs are different. But I'm not really interested in developing against or near a known bug at this point. I'll shelve this as an issue, probably blocked on CRDB #70731.